### PR TITLE
Port moveit ros control interface to ROS2

### PR DESCRIPTION
**Pull request by JafarAbdi**
### Description

- Port moveit_ros_control_interface to ROS2
- trajectory_execution_manager: Multiple bug fixes

Currently setting `moveit_manage_controllers` to `True` will not work since controller_manager doesn't report joints names for the inactive controllers

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_plugins/moveit_ros_control_interface/CMakeLists.txt [hunk 1]
```diff
@@ -1,101 +1,85 @@
...
     moveit_core
+    moveit_simple_controller_manager
+    pluginlib
     trajectory_msgs
-  DEPENDS Boost
 )
 
-include_directories(include)
-include_directories(SYSTEM ${catkin_INCLUDE_DIRS} ${Boost_INCLUDE_DIRS})
+foreach(dependency IN ITEMS ${THIS_PACKAGE_INCLUDE_DEPENDS})
```
Comment (henningkayser) [inline moveit_plugins/moveit_ros_control_interface/CMakeLists.txt:15]: This is a really nice way of using the include depends. Maybe a best practice we want to apply everywhere?

File: moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp [hunk 2]
```diff
@@ -1765,11 +1767,11 @@ bool TrajectoryExecutionManager::ensureActiveControllers(const std::vector<std::
         for (const std::string& controller_to_activate : controllers_to_activate)
         {
           ControllerInformation& ci = known_controllers_[controller_to_activate];
-          ci.last_update_ = rclcpp::Time();
+          ci.last_update_ = rclcpp::Time(0, 0, RCL_ROS_TIME);
```
Comment (henningkayser) [inline moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp:1770]: Looks like this should be set to `rclcpp::Time::now()`, no?

File: moveit_plugins/moveit_ros_control_interface/moveit_ros_control_interface_plugins.xml [hunk 3]
```diff
@@ -1,18 +1,6 @@
...
-    </class>
-    <class name="effort_controllers/JointTrajectoryController" type="moveit_ros_control_interface::JointTrajectoryControllerAllocator" base_class_type="moveit_ros_control_interface::ControllerHandleAllocator">
-        <description></description>
-    </class>
-    <class name="pos_vel_controllers/JointTrajectoryController" type="moveit_ros_control_interface::JointTrajectoryControllerAllocator" base_class_type="moveit_ros_control_interface::ControllerHandleAllocator">
-        <description></description>
-    </class>
-    <class name="pos_vel_acc_controllers/JointTrajectoryController" type="moveit_ros_control_interface::JointTrajectoryControllerAllocator" base_class_type="moveit_ros_control_interface::ControllerHandleAllocator">
+  <library path="moveit_ros_control_interface_trajectory_plugin">
+    <class name="joint_trajectory_controller/JointTrajectoryController" type="moveit_ros_control_interface::JointTrajectoryControllerAllocator" base_class_type="moveit_ros_control_interface::ControllerHandleAllocator">
```
Comment (henningkayser) [inline moveit_plugins/moveit_ros_control_interface/moveit_ros_control_interface_plugins.xml:3]: for better understanding.. All `JointTrajectoryController` types will use the same plugin in the future?

File: moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp [hunk 4]
```diff
@@ -1071,6 +1071,8 @@ bool TrajectoryExecutionManager::configure(TrajectoryExecutionContext& context,
     // empty trajectories don't need to configure anything
     return true;
   }
+
+  reloadControllerInformation();
```
Comment (henningkayser) [inline moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp:1075]: Why is this needed here?

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp [hunk 5]
```diff
@@ -62,18 +67,32 @@ namespace moveit_ros_control_interface
...
 }
 
-MOVEIT_CLASS_FORWARD(MoveItControllerManager)  // Defines MoveItControllerManagerPtr, ConstPtr, WeakPtr... etc
+/**
+ * \brief Get joint name from resource name reported by ros2_control, since claimed_interfaces return by ros2_control
+ * will have the interface name as suffix joint_name/INTERFACE_TYPE
+ * @param[in] claimed_interface claimed interface as joint_name/INTERFACE_TYPE
+ * @return joint_name part of the /p claimed_interface
+ */
+std::string getJointName(const std::string& claimed_interface)
```
Comment (henningkayser) [inline moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:87]:
```suggestion
std::string parseJointNameFromResource(const std::string& claimed_interface)
```
or something similar

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp [hunk 6]
```diff
@@ -62,18 +67,32 @@ namespace moveit_ros_control_interface
...
+{
+  const auto index = claimed_interface.find('/');
+  if (index == std::string::npos)
+    return claimed_interface;
```
Comment (henningkayser) [inline moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:91]: should this be handled in any way?

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp [hunk 7]
```diff
@@ -204,14 +232,27 @@ class MoveItControllerManager : public moveit_controller_manager::MoveItControll
...
+  {
+    node_ = node;
+    if (!ns_.empty())
+    {
+      if (!node_->has_parameter("ros_control_namespace"))
+        ns_ = node_->declare_parameter<std::string>("ros_control_namespace", "/");
+    }
+    list_controllers_service_ = node_->create_client<controller_manager_msgs::srv::ListControllers>(
+        getAbsName("controller_manager/list_controllers"));
+    switch_controller_service_ = node_->create_client<controller_manager_msgs::srv::SwitchController>(
```
Comment (henningkayser) [inline moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:245]: Do we need to wait for the services here to synchronize with ROS2-control?

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp [hunk 8]
```diff
@@ -338,31 +373,29 @@ class MoveItControllerManager : public moveit_controller_manager::MoveItControll
...
-      if (!ros::service::call(getAbsName("controller_manager/switch_controller"), srv))
+      auto result_future = switch_controller_service_->async_send_request(request);
+      if (result_future.wait_for(std::chrono::duration<double>(SERVICE_CALL_TIMEOUT)) == std::future_status::timeout)
       {
-        ROS_ERROR_STREAM("Could switch controllers at " << ns_);
+        RCLCPP_ERROR_STREAM(LOGGER, "Could switch controllers at " << switch_controller_service_->get_service_name());
       }
       discover(true);
-      return srv.response.ok;
+      return result_future.get()->ok;
```
Comment (henningkayser) [inline moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:398]: In case of a timeout `get()` shouldn't be called anymore since it's using an internal wait call.

**Comment by henningkayser** [comment]
@JafarAbdi doesn't this PR address https://github.com/ros-planning/moveit2/issues/507

File: moveit_plugins/moveit_ros_control_interface/moveit_ros_control_interface_plugins.xml (hunk 3 above)
Comment (JafarAbdi) [reply moveit_plugins/moveit_ros_control_interface/moveit_ros_control_interface_plugins.xml:3]: I'm not 100% sure, but I think currently this's the only supported controller, @destogl what do you think .?

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp (hunk 6 above)
Comment (JafarAbdi) [reply moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:91]: Just in case, do you want me to throw an exception if this happens .?

File: moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp (hunk 4 above)
Comment (JafarAbdi) [reply moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp:1075]: This's related to https://github.com/ros-controls/ros2_control/issues/464 causing `known_controllers_` to have an empty vector of joints for the newly activated controller, calling this will update the joint list

File: moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp (hunk 2 above)
Comment (JafarAbdi) [reply moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp:1770]: I think both should work, this how it was previously

**Review by AndyZe — COMMENTED** [review]
Is there a high-level overview of the difference between these packages somewhere? Frankly, they all sound the same:

- moveit_ros_control_interface:  "ros_control controller manager interface for MoveIt"
- moveit_controller_manager: I see that this is COLCON-ignored, probably will be deleted.
- moveit_simple_controller_manager: "A generic, simple controller manager plugin for MoveIt."

It seems like `moveit_ros_control_interface` and `moveit_simple_controller_manager` should be combined into one package.

**Review by AndyZe — COMMENTED** [review]
I also think a bit of specific documentation about what these packages do would be great. Do they switch controllers when a different move group is used, or ...?

File: moveit_plugins/moveit_ros_control_interface/moveit_ros_control_interface_plugins.xml (hunk 3 above)
Comment (destogl) [reply moveit_plugins/moveit_ros_control_interface/moveit_ros_control_interface_plugins.xml:3]: Currently, there is only one `JointTrajectoryController` which interfaces are configurable. We are planning to add specializations into different namespaces, but the current one will always be used at the end. The specializations basically only limit parameterization possibilities.

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp [hunk 9]
```diff
@@ -62,18 +67,32 @@ namespace moveit_ros_control_interface
  * @param[in] force force timeout
  * @return True if timeout duration was passed
  */
-bool checkTimeout(ros::Time& t, double timeout, bool force = false)
+bool checkTimeout(const rclcpp::Clock::SharedPtr& clock, rclcpp::Time& t, double timeout, bool force = false)
```
Comment (henningkayser) [inline moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:70]: This seems overly complicated and the semantics of `force` is somewhat weird here...
I would replace this function with a single-line check like:
```
static const rclcpp::Duration CONTROLLERS_TIMEOUT = rclcpp::Duration::from_seconds(1.0);

...
void discover(bool forceUpdate = false)
{
  if (forceUpdate || (node_->now() - controllers_stamp_ >= CONTROLLERS_TIMEOUT))
  {
    controllers_stamp_ = node_now();
    ...
  }
}
```

File: moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp (hunk 6 above)
Comment (henningkayser) [reply moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:91]: I think that would make a lot of sense. Especially if we don't have a better way to validate the resource at this point.

File: moveit_plugins/moveit_ros_control_interface/src/gripper_controller_plugin.cpp [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+ * \brief Simple allocator for moveit_simple_controller_manager::FollowJointTrajectoryControllerHandle instances.
+ */
+class GripperControllerAllocator : public ControllerHandleAllocator
+{
+public:
+  moveit_controller_manager::MoveItControllerHandlePtr alloc(const rclcpp::Node::SharedPtr& node,
+                                                             const std::string& name,
+                                                             const std::vector<std::string>& /* resources */) override
+  {
+    return std::make_shared<moveit_simple_controller_manager::GripperControllerHandle>(node, name, "gripper_cmd");
```
Comment (henningkayser) [inline moveit_plugins/moveit_ros_control_interface/src/gripper_controller_plugin.cpp:55]: should "gripper_cmd" be dynamic, or is it always going to be this value?

**Review by henningkayser — COMMENTED** [review]
testing this right now...

**Review by henningkayser — APPROVED** [review]
